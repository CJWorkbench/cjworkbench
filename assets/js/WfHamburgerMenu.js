// "Hamburger" drop down on workflow and workflows page. Fixed contents.

import React from 'react'
import PropTypes from 'prop-types'
import { UncontrolledDropdown, DropdownToggle, DropdownMenu, DropdownItem } from './components/Dropdown'
import ImportModuleFromGitHub from './ImportModuleFromGitHub'
import LocaleSwitcher from './i18n/LocaleSwitcher'
import { Trans, t } from '@lingui/macro'
import { withI18n } from '@lingui/react'

export class WfHamburgerMenu extends React.Component {
  static propTypes = {
    i18n: PropTypes.shape({
      // i18n object injected by LinguiJS withI18n()
      _: PropTypes.func.isRequired
    }),
    api: PropTypes.object, // not required: WorkflowListNavBar doesn't allow import from github
    workflowId: PropTypes.number, // not required: WorkflowListNavBar has no workflow
    user: PropTypes.shape({
      id: PropTypes.number.isRequired
    }) // if null/undefined, user is not logged in
  }

  state = {
    importModalOpen: false
  }

  handleClickOpenImportModal = () => {
    this.setState({ importModalOpen: true })
  }

  closeImportModal = () => {
    this.setState({ importModalOpen: false })
  }

  render () {
    const { api, workflowId, user, i18n } = this.props
    const loggedIn = !!user

    return (
      <>
        <LocaleSwitcher />
        <UncontrolledDropdown>

          <DropdownToggle title={i18n._(t('js.WfHamburgerMenu.toggle.hoverText')`menu`)} className='context-button'>
            <i className='icon-more' />
          </DropdownToggle>

          <DropdownMenu>
            {loggedIn && workflowId ? (
              <>
                <DropdownItem href='/workflows/'><Trans id='js.WfHamburgerMenu.menu.myWorkflows'>My Workflows</Trans></DropdownItem>
                <DropdownItem onClick={this.handleClickOpenImportModal}><Trans id='js.WfHamburgerMenu.menu.importModule'>Import Module</Trans></DropdownItem>
              </>
            ) : (
              <DropdownItem href='//workbenchdata.com'><Trans id='js.WfHamburgerMenu.menu.home'>Home</Trans></DropdownItem>
            )}
            {loggedIn ? (
              <DropdownItem href='/account/logout'><Trans id='js.WfHamburgerMenu.menu.logout'>Log Out</Trans></DropdownItem>
            ) : null}
          </DropdownMenu>
        </UncontrolledDropdown>
        {this.state.importModalOpen ? (
          <ImportModuleFromGitHub
            closeModal={this.closeImportModal}
            api={api}
          />
        ) : null}
      </>
    )
  }
}

export default withI18n()(WfHamburgerMenu)
