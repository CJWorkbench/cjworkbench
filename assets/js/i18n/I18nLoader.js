import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
