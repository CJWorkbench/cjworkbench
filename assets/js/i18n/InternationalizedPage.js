import React from 'react'
import PropTypes from 'prop-types'
import { en, el } from 'make-plural/plurals'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { currentLocaleId } from './locales'
import Spinner from '../Spinner'

__webpack_public_path__ = window.STATIC_URL + 'bundles/' // eslint-disable-line

async function loadLocale (localeId) {
  const { messages } = await import(
    /* webpackChunkName: "[request]" */
    `@lingui/loader!../../locale/${localeId}/messages.po`
  )
  i18n.loadLocaleData({
    en: { plurals: en },
    el: { plurals: el }
  })
  i18n.load(localeId, messages)
  i18n.activate(localeId)
  return i18n
}

function useActivatedI18n (localeId) {
  const [i18n, setI18n] = React.useState(null)

  React.useEffect(() => {
    loadLocale(currentLocaleId).then(setI18n)
  }, [])

  return i18n
}

export default function InternationalizedPage (props) {
  const { children } = props
  const i18n = useActivatedI18n(currentLocaleId)

  if (i18n === null) {
    return <Spinner />
  }

  return (
    <I18nProvider i18n={i18n} forceRenderOnLocaleChange={false}>
      {children}
    </I18nProvider>
  )
}
InternationalizedPage.propTypes = {
  children: PropTypes.node.isRequired
}
