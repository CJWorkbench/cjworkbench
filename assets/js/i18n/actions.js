const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
