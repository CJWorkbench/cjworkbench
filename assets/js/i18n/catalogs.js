import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
