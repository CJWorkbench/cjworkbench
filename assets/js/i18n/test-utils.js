import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
