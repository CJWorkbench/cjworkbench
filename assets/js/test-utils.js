/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
