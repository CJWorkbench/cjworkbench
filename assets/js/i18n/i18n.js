import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
