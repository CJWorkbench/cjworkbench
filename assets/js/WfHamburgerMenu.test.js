/* globals afterEach, describe, expect, it */
import React from 'react'
import { WfHamburgerMenu } from './WfHamburgerMenu'
import { mountWithI18n } from './i18n/test-utils'

describe('WfHamburgerMenu', () => {
  let wrapper // all tests must mount one
  afterEach(() => wrapper.unmount())

  it('renders logged in, non-read only', () => {
    wrapper = mountWithI18n(
      <WfHamburgerMenu
        workflowId={1}
        api={{}}
        isReadOnly={false}
        user={{ id: 100 }}
      />
    )

    wrapper.find('button.context-button').simulate('click')
    expect(wrapper).toMatchSnapshot() // one snapshot only, in most common case

    expect(wrapper.find('a[href="/workflows/"]')).toHaveLength(1)
    expect(wrapper.find('DropdownItem Trans[defaults="Import Module"]')).toHaveLength(1)
    expect(wrapper.find('DropdownItem Trans[defaults="Log Out"]')).toHaveLength(1)
  })

  it('renders logged in, read only', () => {
    wrapper = mountWithI18n(
      <WfHamburgerMenu
        workflowId={1}
        api={{}}
        isReadOnly
        user={{ id: 100 }}
      />
    )

    wrapper.find('button.context-button').simulate('click')
    expect(wrapper.find('a[href="/workflows/"]')).toHaveLength(1)
    expect(wrapper.find('DropdownItem Trans[defaults="Import Module"]')).toHaveLength(1)
    expect(wrapper.find('DropdownItem Trans[defaults="Log Out"]')).toHaveLength(1)
  })

  it('renders logged out, read only', () => {
    wrapper = mountWithI18n(
      <WfHamburgerMenu
        workflowId={1}
        api={{}}
        isReadOnly
        user={undefined}
      />
    )

    wrapper.find('button.context-button').simulate('click')
    expect(wrapper.find('a[href="//workbenchdata.com"]')).toHaveLength(1)
    expect(wrapper.find('DropdownItem[children="Import Module"]')).toHaveLength(0)
    expect(wrapper.find('DropdownItem[children="Log out"]')).toHaveLength(0)
  })

  it('renders without a workflowId', () => {
    // this happens on Workflow list page
    wrapper = mountWithI18n(
      <WfHamburgerMenu
        api={{}}
        isReadOnly
        user={{ id: 100 }}
      />
    )

    wrapper.find('button.context-button').simulate('click')
    expect(wrapper.find('a[href="//workbenchdata.com"]')).toHaveLength(1)
    expect(wrapper.find('DropdownItem[children="Import Module"]')).toHaveLength(0)
    expect(wrapper.find('DropdownItem Trans[defaults="Log Out"]')).toHaveLength(1)
  })
})
