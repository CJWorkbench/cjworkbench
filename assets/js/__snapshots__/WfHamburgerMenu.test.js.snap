// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WfHamburgerMenu renders logged in, non-read only 1`] = `
<WfHamburgerMenu
  api={Object {}}
  i18n={
    Object {
      "_": [Function],
      "_activeLanguageData": Object {
        "plurals": [Function],
      },
      "_activeMessages": Object {},
      "_catalogs": Object {
        "en": Object {
          "languageData": Object {},
          "messages": Object {},
        },
      },
      "_dev": Object {
        "compile": [Function],
        "loadLanguageData": [Function],
      },
      "_language": "en",
      "_locales": undefined,
      "_missing": undefined,
      "plural": [Function],
      "select": [Function],
      "selectOrdinal": [Function],
      "t": [Function],
    }
  }
  isReadOnly={false}
  user={
    Object {
      "id": 100,
    }
  }
  workflowId={1}
>
  <WithI18n>
    <I18n
      update={true}
      withHash={true}
    >
      <Memo(LocaleSwitcher)
        i18n={
          Object {
            "_": [Function],
            "_activeLanguageData": Object {
              "plurals": [Function],
            },
            "_activeMessages": Object {},
            "_catalogs": Object {
              "en": Object {
                "languageData": Object {},
                "messages": Object {},
              },
            },
            "_dev": Object {
              "compile": [Function],
              "loadLanguageData": [Function],
            },
            "_language": "en",
            "_locales": undefined,
            "_missing": undefined,
            "plural": [Function],
            "select": [Function],
            "selectOrdinal": [Function],
            "t": [Function],
          }
        }
      />
    </I18n>
  </WithI18n>
  <UncontrolledDropdown>
    <Dropdown
      isOpen={true}
      toggle={[Function]}
    >
      <Manager
        tag={false}
      >
        <div
          className="dropdown"
        >
          <DropdownToggle
            className="context-button"
            title="menu"
          >
            <Reference>
              <InnerReference
                setReferenceNode={[Function]}
              >
                <button
                  aria-expanded={true}
                  className="btn btn-secondary context-button"
                  disabled={false}
                  onClick={[Function]}
                  title="menu"
                >
                  <i
                    className="icon-more"
                  />
                </button>
              </InnerReference>
            </Reference>
          </DropdownToggle>
          <DropdownMenu>
            <OpenDropdownMenu>
              <Portal
                containerInfo={
                  <body>
                    <div
                      class="dropdown-menu-portal"
                      style="position: absolute; top: 0px; left: 0px; opacity: 0; pointer-events: none;"
                    >
                      <div
                        class="dropdown-menu"
                        role="menu"
                      >
                        <a
                          class="dropdown-item"
                          href="/workflows/"
                          role="menuitem"
                          tabindex="0"
                        >
                          js.WfHamburgerMenu.menu.myWorkflows
                        </a>
                        <button
                          class="dropdown-item"
                          role="menuitem"
                          tabindex="0"
                        >
                          js.WfHamburgerMenu.menu.importModule
                        </button>
                        <a
                          class="dropdown-item"
                          href="/account/logout"
                          role="menuitem"
                          tabindex="0"
                        >
                          js.WfHamburgerMenu.menu.logout
                        </a>
                      </div>
                    </div>
                  </body>
                }
              >
                <Popper
                  modifiers={
                    Object {
                      "preventOverflow": Object {
                        "boundariesElement": "viewport",
                      },
                    }
                  }
                  placement="bottom-end"
                >
                  <InnerPopper
                    eventsEnabled={true}
                    modifiers={
                      Object {
                        "preventOverflow": Object {
                          "boundariesElement": "viewport",
                        },
                      }
                    }
                    placement="bottom-end"
                    positionFixed={false}
                    referenceElement={
                      <button
                        aria-expanded="true"
                        class="btn btn-secondary context-button"
                        title="menu"
                      >
                        <i
                          class="icon-more"
                        />
                      </button>
                    }
                  >
                    <div
                      className="dropdown-menu-portal"
                      style={
                        Object {
                          "left": 0,
                          "opacity": 0,
                          "pointerEvents": "none",
                          "position": "absolute",
                          "top": 0,
                        }
                      }
                    >
                      <div
                        className="dropdown-menu"
                        role="menu"
                      >
                        <DropdownItem
                          href="/workflows/"
                        >
                          <a
                            className="dropdown-item"
                            href="/workflows/"
                            role="menuitem"
                            tabIndex={0}
                          >
                            <WithI18n
                              defaults="My Workflows"
                              id="js.WfHamburgerMenu.menu.myWorkflows"
                            >
                              <I18n
                                update={true}
                                withHash={true}
                              >
                                <Trans
                                  defaults="My Workflows"
                                  i18n={
                                    Object {
                                      "_": [Function],
                                      "_activeLanguageData": Object {
                                        "plurals": [Function],
                                      },
                                      "_activeMessages": Object {},
                                      "_catalogs": Object {
                                        "en": Object {
                                          "languageData": Object {},
                                          "messages": Object {},
                                        },
                                      },
                                      "_dev": Object {
                                        "compile": [Function],
                                        "loadLanguageData": [Function],
                                      },
                                      "_language": "en",
                                      "_locales": undefined,
                                      "_missing": undefined,
                                      "plural": [Function],
                                      "select": [Function],
                                      "selectOrdinal": [Function],
                                      "t": [Function],
                                    }
                                  }
                                  id="js.WfHamburgerMenu.menu.myWorkflows"
                                >
                                  <Render
                                    value="js.WfHamburgerMenu.menu.myWorkflows"
                                  >
                                    js.WfHamburgerMenu.menu.myWorkflows
                                  </Render>
                                </Trans>
                              </I18n>
                            </WithI18n>
                          </a>
                        </DropdownItem>
                        <DropdownItem
                          onClick={[Function]}
                        >
                          <button
                            className="dropdown-item"
                            onClick={[Function]}
                            role="menuitem"
                            tabIndex={0}
                          >
                            <WithI18n
                              defaults="Import Module"
                              id="js.WfHamburgerMenu.menu.importModule"
                            >
                              <I18n
                                update={true}
                                withHash={true}
                              >
                                <Trans
                                  defaults="Import Module"
                                  i18n={
                                    Object {
                                      "_": [Function],
                                      "_activeLanguageData": Object {
                                        "plurals": [Function],
                                      },
                                      "_activeMessages": Object {},
                                      "_catalogs": Object {
                                        "en": Object {
                                          "languageData": Object {},
                                          "messages": Object {},
                                        },
                                      },
                                      "_dev": Object {
                                        "compile": [Function],
                                        "loadLanguageData": [Function],
                                      },
                                      "_language": "en",
                                      "_locales": undefined,
                                      "_missing": undefined,
                                      "plural": [Function],
                                      "select": [Function],
                                      "selectOrdinal": [Function],
                                      "t": [Function],
                                    }
                                  }
                                  id="js.WfHamburgerMenu.menu.importModule"
                                >
                                  <Render
                                    value="js.WfHamburgerMenu.menu.importModule"
                                  >
                                    js.WfHamburgerMenu.menu.importModule
                                  </Render>
                                </Trans>
                              </I18n>
                            </WithI18n>
                          </button>
                        </DropdownItem>
                        <DropdownItem
                          href="/account/logout"
                        >
                          <a
                            className="dropdown-item"
                            href="/account/logout"
                            role="menuitem"
                            tabIndex={0}
                          >
                            <WithI18n
                              defaults="Log Out"
                              id="js.WfHamburgerMenu.menu.logout"
                            >
                              <I18n
                                update={true}
                                withHash={true}
                              >
                                <Trans
                                  defaults="Log Out"
                                  i18n={
                                    Object {
                                      "_": [Function],
                                      "_activeLanguageData": Object {
                                        "plurals": [Function],
                                      },
                                      "_activeMessages": Object {},
                                      "_catalogs": Object {
                                        "en": Object {
                                          "languageData": Object {},
                                          "messages": Object {},
                                        },
                                      },
                                      "_dev": Object {
                                        "compile": [Function],
                                        "loadLanguageData": [Function],
                                      },
                                      "_language": "en",
                                      "_locales": undefined,
                                      "_missing": undefined,
                                      "plural": [Function],
                                      "select": [Function],
                                      "selectOrdinal": [Function],
                                      "t": [Function],
                                    }
                                  }
                                  id="js.WfHamburgerMenu.menu.logout"
                                >
                                  <Render
                                    value="js.WfHamburgerMenu.menu.logout"
                                  >
                                    js.WfHamburgerMenu.menu.logout
                                  </Render>
                                </Trans>
                              </I18n>
                            </WithI18n>
                          </a>
                        </DropdownItem>
                      </div>
                    </div>
                  </InnerPopper>
                </Popper>
              </Portal>
            </OpenDropdownMenu>
          </DropdownMenu>
        </div>
      </Manager>
    </Dropdown>
  </UncontrolledDropdown>
</WfHamburgerMenu>
`;
